text/0000-monorepo-support.md

```diff
@@ -0,0 +1,81 @@
+
+Importantly, the order of the sequence of project roots will start with the closest `package.json` (what we previously called the sole project root), and then follow the `"extends"` chain out from there.
+
+## Platform Resolution
+
+To resolve the Project platform, taking into account all of the settings in the different project files, we will merge the platform settings from each project file, giving precedence at all times to the files earlier in the sequence. So the highest precedence will be at the existing project root, and then each file that we resolve after that will only update settings that haven't been specified by an earlier entry.
+
+To support a use-case where a project wants to exclude e.g. `yarn` and _not_ allow a later project file to add one, we will support a new value for the package manager keys in the `"volta"` section: `false`. Setting `"yarn": false` will mean that no Yarn is available _and_ that a file later in the sequence with a `yarn` setting won't be able to override that removal of Yarn.
+
+Additionally, while we are processing this chain of project files, we will need to take care to detect accidental loops in the files pointed to by `"extends"` and exit early with a helpful error message (as opposed to entering an infinite loop of parsing and never finishing).
+
+Finally, in lieu of the current parse error for not setting `"node"` in the `"volta"` section, we should merge all of the project settings into a single project Platform, and then only show an error if none of the files specified a Node version.
+
+## Local Binary Resolution
+
+Currently, in order to run a project-local binary, we look for the executable in `<project root>/node_modules/.bin/`. With multiple project roots, this process will be almost the same: For each root in the sequence, we check for the binary in the expected place, and if it's found, then we stop and use that version. If we don't find it, we move on to the next root in the sequence.
+
+## Project-Local Hooks
+
+Similar to how we resolve the Platform, we will resolve project-local hooks by looking for `<project root>/.volta/hooks.json` for each project root. We will merge them together, with earlier entries having precedence over later ones, and then finally merge the result with the user-level hooks in the same way that we already do to merge project and user-level hooks.
+
+## Pinning Tools
+
+When a user runs `volta pin`, we will always make the changes to the _first_ `package.json` that we find. This means that if a user wants to `pin` a tool for the root project, they will need to first navigate to the root directory (out of any subprojects) and _then_ run `volta pin`. This gives users the most flexibility to control where they want to `pin` tools, without trying to guess which location they were intending or requiring interactivity.
```

Comment:
Can we make the behavior of `volta pin` workspaces aware? Most of the time when you are working in a monorepo you are running your terminal in the top level project root. Assuming that is the case, then it seems like we could make `volta pin` notice that there is a `workspaces` property in that `package.json` and from that we can update all of the `package.json` files for you (with the right `volta: { extends: '../../package.json' }` bit.

Having to do it manually is fine (and the result with `extends` proposed here is much better), but IMHO we can still do **exactly** what the user wants 98% of the time...

Reply:
I agree we want to do what the user wants, but I'm also a little wary of a command that normally only affects a single `package.json` suddenly having knock-on effects in a bunch of different files. My concern is that if someone wants to pin the top-level _without_ updating the subprojects to have `extends`, then we don't really give a way to do that (apart from manual edits). Either that or we have to add a flag to the CLI that only applies if you are A) In the root of a Yarn Workspaces project and B) Trying to pin the top-level without making any changes to subprojects, which is a little to specific to be useful as a command-line flag, IMO.

That said, the benefits to getting started with a workspace (where we know we have a reverse mapping to the individual projects) are very nice, so it feels like it would be a good QoL improvement. I'll add it and we'll see if there are other comments / issues raised about how it works.